controller/internal/enforcer/envoyauthorizer/envoyauthorizerenforcer.go

```diff
@@ -82,16 +97,195 @@ func (e *Enforcer) Secrets() (secrets.Secrets, func()) {
 }
 
 // Enforce starts enforcing policies for the given policy.PUInfo.
+// here we do the following:
+// 1. create a new PU always and instantiate a new apiAuth, as we want to be as stateless as possible.
+// 2. create a PUcontext as this will be used in auth code.
+// 3. If envoy servers are not present then create all 3 envoy servers.
+// 4. If the servers are already present under policy update then update the service certs.
 func (e *Enforcer) Enforce(contextID string, puInfo *policy.PUInfo) error {
 	e.Lock()
 	defer e.Unlock()
+
+	zap.L().Debug("Enforce for the envoy for pu: ", zap.String("puID:", contextID))
+	// here we 1st need to create a PuContext, as the PU context will derive the
+	// serviceCtxt which will be used by the authorizer to determine the policyInfo.
+
+	pu, err := pucontext.NewPU(contextID, puInfo, e.externalIPCacheTimeout)
+	if err != nil {
+		return fmt.Errorf("error creating new pu: %s", err)
+	}
+	// Add the puContext to the cache as we need to later while serving the requests.
+	e.puContexts.AddOrUpdate(contextID, pu)
+
+	sctx, err := e.registry.Register(contextID, puInfo, pu, e.secrets)
+	if err != nil {
+		return fmt.Errorf("policy conflicts detected: %s", err)
+	}
+
+	caPool := e.expandCAPool(sctx.RootCA)
+
+	// now instantiate the apiAuth and metadata
+	// create a new server if it doesn't exist yet
+	if _, err := e.clients.Get(contextID); err != nil {
+		zap.L().Debug("creating new auth and sds servers", zap.String("puID", contextID))
+		ingressServer, err := envoyproxy.NewExtAuthzServer(contextID, e.puContexts, e.collector, envoyproxy.IngressDirection, e.registry, e.secrets, e.tokenIssuer)
+		if err != nil {
+			zap.L().Error("Cannot create and run IngressServer", zap.Error(err))
+			return err
+		}
+
+		egressServer, err := envoyproxy.NewExtAuthzServer(contextID, e.puContexts, e.collector, envoyproxy.EgressDirection, e.registry, e.secrets, e.tokenIssuer)
+		if err != nil {
+			zap.L().Error("Cannot create and run EgressServer", zap.Error(err))
+			ingressServer.Stop()
+			return err
+		}
+		sdsServer, err := envoyproxy.NewSdsServer(contextID, puInfo, caPool, e.secrets)
+		if err != nil {
+			zap.L().Error("Cannot create and run SdsServer", zap.Error(err))
+			return err
```

Comment:
Do you need to stop the ingressServer and egressServer before returning ?
```
ingressServer.Stop()
egressServer.Stop()
return err
```

Reply:
looks correct yes

Reply:
My Bad, good catch @brianonn  , will make the changes and commit to master.